rust/src/io/commit.rs

```diff
@@ -55,35 +58,98 @@ pub type ManifestWriter = for<'a> fn(
     path: &'a Path,
 ) -> BoxFuture<'a, Result<()>>;
 
+const LATEST_MANIFEST_NAME: &str = "_latest.manifest";
+const VERSIONS_DIR: &str = "_versions";
+
+/// Get the manifest file path for a version.
+fn manifest_path(base: &Path, version: u64) -> Path {
+    base.child(VERSIONS_DIR)
+        .child(format!("{version}.manifest"))
+}
+
+/// Get the latest manifest path
+fn latest_manifest_path(base: &Path) -> Path {
+    base.child(LATEST_MANIFEST_NAME)
+}
+
+async fn list_manifests<'a>(
+    base_path: &Path,
+    object_store: &'a ObjectStore,
+) -> Result<BoxStream<'a, Result<Path>>> {
+    let base_path = base_path.clone();
+    Ok(object_store
+        .read_dir_all(&base_path.child(VERSIONS_DIR), None)
+        .await?
+        .try_filter(|path| future::ready(path.extension() == Some("manifest")))
+        .boxed())
+}
+
+async fn write_latest_manifest(
+    commit_handler: &dyn CommitHandler,
+    from_path: &Path,
+    base_path: &Path,
+    object_store: &ObjectStore,
+) -> Result<()> {
+    let latest_path = commit_handler
+        .resolve_latest_version(base_path, object_store)
+        .await?;
+    // Then, create a copy at the latest manifest path
+    object_store
+        .inner
+        .copy(from_path, &latest_path)
+        .await
+        .map_err(|err| CommitError::OtherError(err.into()))?;
+    Ok(())
+}
+
 /// Handle commits that prevent conflicting writes.
 ///
 /// Commit implementations ensure that if there are multiple concurrent writers
 /// attempting to write the next version of a table, only one will win. In order
 /// to work, all writers must use the same commit handler type.
+/// This trait is also responsible for
 #[async_trait::async_trait]
 pub trait CommitHandler: Debug + Send + Sync {
-    /// Get the latest version of a dataset at the path
-    async fn get_latest_version(
+    /// Get the path to the latest version manifest of a dataset at the base_path
+    async fn resolve_latest_version(
         &self,
         path: &Path,
-        object_store: &ObjectStore,
-    ) -> std::result::Result<u64, crate::Error> {
+        _object_store: &ObjectStore,
+    ) -> std::result::Result<Path, crate::Error> {
         // use the _latest.manifest file to get the latest version
         // TODO: this isn't 100% safe, we should list the /_versions directory and find the latest version
         // TODO: we need to pade 0's to the version number on the manifest file path
-        read_manifest(object_store, &latest_manifest_path(path))
-            .await
-            .map(|m| m.version)
+        Ok(latest_manifest_path(path))
+    }
+
+    /// Get the path to a specific versioned manifest of a dataset at the base_path
+    async fn resolve_version(
```

Comment:
Part of me is nervous about opening up all of this for extension. IIUC, this would allow having an external store be the manifest store and completely bypass the files in the table. Not sure we want to do that right now.

What would we think of exposing these as hooks?

```rust
async fn before_get_manifest(
    &self,
    base_path: &Path,
    version: u64,
    object_store: &ObjectStore,
) -> Result<()> {};
```

Then no matter what, `resolve_version` will look in the same place for a manifest file. It's just that for some backends, you can provide a hook to sync things if needed.

Reply:
Since `CommitHandler` is still private to the crate, I think this is okay since we control all the implementations. We can decide if we want to expose hooks or the entire thing later. If we want to expose just hooks we can make another trait for it.

Reply:
Actually, it is public:
https://docs.rs/lance/0.7.0/lance/io/commit/trait.CommitHandler.html

Reply:
hmmm, yeah I didn't mean to expose the entire interface. Let me split them up and keep the old commit handler public

Reply:
ended up taking this interface private as `pub(crate)`